src/compiler/checker.ts

```diff
@@ -20929,8 +20929,7 @@ namespace ts {
                     return type.symbol;
                 }
                 if (isTupleType(type)) {
-                    // Tuple types are tracked through their target type
-                    return type.target;
+                    return type;

```

Comment:
I'm not overly confident that it's the correct **shape** of this fix but all tests pass. 

The problem was that for the inner tuple the same recursion identity was computed here as the `type.target` was shared between those tuples (as they have the same shape). And thus `invokeOnce` was skipping over `action(source, target)` call there while inferring for the inner tuple.

This check is done within this if statement:
https://github.dev/microsoft/TypeScript/blob/942d2a91c115f22857513a9d3f4bde2fb1521f50/src/compiler/checker.ts#L20918-L20919

However, this still goes into this if statement for a tuple because it doesn't satisfy the `isObjectOrArrayLiteralType(type)` check. I think that perhaps a better fix here would be to make a change to this if condition. I've tried this alternative patch and all tests seem to be passing to:
<details>
<summary>Alternative patch</summary>

```diff
diff --git a/src/compiler/checker.ts b/src/compiler/checker.ts
index 7bcc6f5bdc..efce2901b4 100644
--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -20916,7 +20916,7 @@ namespace ts {
         // reference the type have a recursion identity that differs from the object identity.
         function getRecursionIdentity(type: Type): object {
             // Object and array literals are known not to contain recursive references and don't need a recursion identity.
-            if (type.flags & TypeFlags.Object && !isObjectOrArrayLiteralType(type)) {
+            if (type.flags & TypeFlags.Object && !isObjectOrArrayLiteralType(type) && !isTupleType(type)) {
                 if (getObjectFlags(type) && ObjectFlags.Reference && (type as TypeReference).node) {
                     // Deferred type references are tracked through their associated AST node. This gives us finer
                     // granularity than using their associated target because each manifest type reference has a
@@ -20928,10 +20928,6 @@ namespace ts {
                     // exclude the static side of classes from this check since it shares its symbol with the instance side.
                     return type.symbol;
                 }
-                if (isTupleType(type)) {
-                    // Tuple types are tracked through their target type
-                    return type.target;
-                }
             }
             if (type.flags & TypeFlags.TypeParameter) {
                 return type.symbol;
```
</details>

The comment above the `isTupleType` suggests that it's checking if the type was written as literal tuple type:
https://github.dev/microsoft/TypeScript/blob/942d2a91c115f22857513a9d3f4bde2fb1521f50/src/compiler/checker.ts#L21278-L21282
and the comment above the mentioned condition suggests that we don't want to go into this block if a type was written as literal (it only mentions arrays and objects though):
https://github.dev/microsoft/TypeScript/blob/942d2a91c115f22857513a9d3f4bde2fb1521f50/src/compiler/checker.ts#L20918

So perhaps this `isTupleType` check shouldn't be inside this if block from the start?